### ISSUE TITLE
Site become unresponsive after activate the plugin

### ISSUE THREAD

**Issue by vincent-lu** [lcache/wp-lcache#137]
Hi first of all thanks for open source this plugin.

I could use some help with troubleshooting activating this plugin on our WooCommerce site (nginx, php7.0, mysql 5.7, everything on AWS).

I've installed `wp-lcache` according to the installation guide (e.g. installed `apcu`, created `object-cache.php` file, activated the plugin). From what I can see the plugin is definitely running (no error message in backend).

Shortly after I've done above (3~5min), the website become unresponsive, even with low to moderate public traffic. Here are my observations when this happened:

- CPU load on web server (on php-fpm processes) is high but not too high to handle the traffic
- CPU load on mysql server (I'm on Amazon RDS, mysql 5.7) is stable to low
- DB connections on mysql server shoot thru the roof (connection increased from 3~5 to 200, probably capped by RDS itself)
- I can connect to the database via command line fine, the `wp_lcache_tags` table is empty, and `wp_lcache_events` table has around 37k rows in about 5min after activated the plugin
- Here is the screenshot of the database server load around the time I tried to activate the plugin:
![Image of the database server load around the time I tried to activate the plugin](http://i.imgur.com/29w1OaY.png)

I rebooted the database server, again the db connection shoot up (this time capped at 100) but still unresponsive.

I removed the `object-cache.php` and rebooted both web and db server, everything went back to normal.

With redis backed object caching plugins having various problems, I'd really like to take the advantage that this plugin offers, but I'm not sure what exactly is causing this problem. My best guess is `wp-lcache` tries to "warm up" its cache table but due to the amount of data we have (again, WooCommerce site, with about 4000+ products, 40k+ orders, 25k+ media files and negligible amount of posts/pages), the plugin caused some problem with database and made the whole website unresponsive.

Any suggestions on how to pin down the problem?

**Comment by danielbachhuber** [comment]
Hey @vincent-lu,

Thanks for the detailed report. Can you review #138 (which I just posted)? The symptoms sound similar to what you've described.

> With redis backed object caching plugins having various problems,

Out of curiosity, have you tried [WP Redis](https://github.com/pantheon-systems/wp-redis)? The only known issue I'm aware of is [this one](https://github.com/pantheon-systems/wp-redis#how-do-i-disable-the-persistent-object-cache-for-a-bad-actor), so it should meet your needs as long as your code is optimized for the purpose.

**Comment by vincent-lu** [comment]
Hi @danielbachhuber thanks for getting back to me.

I will test out the script in the next few days. But it describes "php processes lock up" which is different from what I observed (where php itself is fine but it's mysql connection went crazy). Regardless, I will give it a go and see if:
- the script can reproduce the lock-up described on our staging environment
- the php lock-up it mentioned share any similarity with the problem we encountered

Regarding Redis based solution, I've been toying with https://github.com/tillkruss/redis-cache and I was able to very occasionally observe the problem described here https://github.com/tillkruss/redis-cache/issues/58 (caused by a race condition bug in WP core). Which make me extremely reluctant to try `tillkruss/redis-cache` on our production. And assuming it's a bug derived from WP core, I'm worried that `pantheon-systems/wp-redis` may suffer similar problems?

Either way because our current WP setup is chewing up a lot CPU circles on our database server (4000+ queries just to list last 200 orders in WooCommerce backend, argh...), I'll have to explore the Redis based solutions more.

**Comment by danielbachhuber** [comment]
> But it describes "php processes lock up" which is different from what I observed (where php itself is fine but it's mysql connection went crazy).

Makes sense. Database usage is another potential performance vector because WP LCache does write to the database. If your WordPress instance isn't very cache performant, meaning there is a high percentage of cache writes happening, then you'll see that manifest in database utilization.

> And assuming it's a bug derived from WP core, I'm worried that pantheon-systems/wp-redis may suffer similar problems?

It does, see https://github.com/pantheon-systems/wp-redis/pull/157#issuecomment-266444418

The workaround from WordPress.com VIP seems like it would be sufficient though: https://github.com/Automattic/vip-mu-plugins-public/blob/master/misc.php#L66-L83

**Comment by vincent-lu** [comment]
> Out of curiosity, have you tried WP Redis? The only known issue I'm aware of is this one, **so it should meet your needs as long as your code is optimized for the purpose.**

@danielbachhuber when you say "as long as your code is optimized for the purpose", do you mean the php code shouldn't use too many object caches? It'd be really hard to manage this side as just with a single plugin activated (WooCommerce), viewing 200 orders on backend results in 30000+ object caches (88.3% hit rate (27,422 hits, 3,628 misses), according to Query Monitor).

**Comment by danielbachhuber** [comment]
> do you mean the php code shouldn't use too many object caches? It'd be really hard to manage this side as just with a single plugin activated (WooCommerce), viewing 200 orders on backend results in 30000+ object caches (88.3% hit rate (27,422 hits, 3,628 misses), according to Query Monitor).

The PHP code can use as many object caches as it can, as long as there's a sufficient hit rate.

If the miss rate is quite high (which it seems like yours is), then it defeats the purpose of the object cache, as the object cache isn't being used much. Furthermore, with LCache, the issue is compounded because cache writes to the L2 incur database writes.

**Comment by vincent-lu** [comment]
The hit rate I saw after I enabled lcache is very high (95%+) consistently. It's really the database problem that's the show stopper and there doesn't seem to be a solution. My guess is the L2 caching system is using too much db resources (37k row in about 5min on `wp_lcache_events ` table does sounds like a lot).

Also for #138 , I tried to run it on our production during offpeak hours, and the symptom is completely different than what I experienced after activating lcache - all it did is shoot in CPU process to 100% nothing else.

For L2 cache, does lcache generates `wp_lcache_events` records right/soon after activation? Does lcache try to throttle itself in L2 generation? Also all our records in `wp_lcache_events` table's pool column is `10.0.0.1:443`, is there something wrong with it using internal ip?

**Comment by danielbachhuber** [comment]
> The hit rate I saw after I enabled lcache is very high (95%+) consistently. It's really the database problem that's the show stopper and there doesn't seem to be a solution. My guess is the L2 caching system is using too much db resources (37k row in about 5min on `wp_lcache_events` table does sounds like a lot).

It's high, but it may not be high enough. Given your total number of cache entries, it seems like the 5% is sufficient to cause ongoing database load issues.

> Also for #138 , I tried to run it on our production during offpeak hours, and the symptom is completely different than what I experienced after activating lcache - all it did is shoot in CPU process to 100% nothing else.

Interesting. Thanks for trying it out.

> For L2 cache, does lcache generates `wp_lcache_events` records right/soon after activation?

Yes. The `wp_lcache_events` records represent all data in the object cache. As soon as LCache is responsible for the object cache, it will begin to populate the L2 cache.

> Does lcache try to throttle itself in L2 generation?

Not at this point, no.

> Also all our records in `wp_lcache_events` table's pool column is 10.0.0.1:443, is there something wrong with it using internal ip?

There shouldn't be. This is just for record-keeping purpose — the data isn't ever publicly exposed.

I'm going to spend some time this month working on a guide for WooCommerce performance optimization. Would you be interested / willing to provide some input based on your experiences?

**Comment by vincent-lu** [comment]
>> Also for #138 , I tried to run it on our production during offpeak hours, and the symptom is completely different than what I experienced after activating lcache - all it did is shoot in CPU process to 100% nothing else.

> Interesting. Thanks for trying it out.

It makes perfect sense though, considering that script is just using apcu and not doing database queries anyway.

>> Does lcache try to throttle itself in L2 generation?

> Not at this point, no.

Yeah I think some form of throttling might help with the problem I ran into; if one day you guys get around to do it.

> I'm going to spend some time this month working on a guide for WooCommerce performance optimization. Would you be interested / willing to provide some input based on your experiences?

Yeah sure. Just let me know in what way you think I can contribute.

Now I've activated https://github.com/pantheon-systems/wp-redis (along with the alloptions fix) on our production (1 web server and redis is also running on it, and 1 db server, wp-redis object caching plus nginx page caching backed by the same redis service) for about 24 hours so far. Here are the findings:

- The web server's CPU load jumped about 20%
- The memory usage increased mildly on the web server
- The network activity (both in and out) on the web server dropped significantly (about 5~10 times)
- The CPU load on database server pretty much stayed as is (to my disappointment), maybe slightly less spiky but not particularly noticeable
- Object cache hit rate is good, on home page it is: Cache Hits:30093 / Cache Misses:502 / Redis Calls:1516: / - get: 1515 /- set: 1
- The page load speed increased and database query time reduced significantly for both frontend and backend

Overall the redis object cache does improve performance in load speed, but not much reduction on resource usages.

I will close this issue for now; thanks @danielbachhuber for all the fish.
